torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -287,13 +304,139 @@ class TensorExprFuser {
   TensorExprFuser(std::shared_ptr<Graph> graph, size_t min_group_size)
       : graph_(std::move(graph)), min_group_size_(min_group_size) {}
 
+  // Builds up expressions that compute shapes of all intermediates (and
+  // outputs) of the fusion group, based on the sizes of inputs. You should run
+  // DCE to remove those that you end up not using.
+  std::unordered_map<Value*, Value*> buildShapeExpressions(Node* fusion_group) {
```

Comment:
How much of this is copypasta from graph_fuser's optimization pass?  Is it similar enough that the logic could be factored out into a common utility?  Or are there some subtleties that would make that messy?

Reply:
`buildShapeExpressions` looks different enough due to the difference in a way we are handling `aten::cat`.  In TE fusion groups, it's possible to have a `aten::cat` to be stuck in the middle of a group vs  to only appear as an output in OF's fusion groups. I needed to add logic to detect this scenario and bailout. I did just notice we could share `removeOutputsUsedOnlyInSize` though